### PR TITLE
[12.1] ISPN-12636 Implicit connectors

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-12636

### CODE REVIEW

File: documentation/src/main/asciidoc/topics/ref_endpoints_connectors.adoc [hunk 1]
```diff
@@ -18,6 +18,8 @@ and whether clients must present valid TLS certificates.
 * `<memcached-connector socket-binding="memcached"/>` declares a Memcached
 connector that uses the memcached socket binding.
 
+Declaring an empty `<endpoints/>` element implicitly enables the Hot Rod and REST connectors.
```
Comment (ryanemerson) [inline documentation/src/main/asciidoc/topics/ref_endpoints_connectors.adoc:21]: To disable the Hotrod endpoint do I need todo the following?

```
<endpoints>
  <rest-connector />
</endpoints>
```

Comment (tristantarrant) [reply documentation/src/main/asciidoc/topics/ref_endpoints_connectors.adoc:21]: yes

**Review by pruivo — CHANGES_REQUESTED** [review]
I'm unable to access the console with implicit connectors. Steps I did:

* compile this PR
* `cd server/runtime/target/infinispan-server-12.1.0-SNAPSHOT`
* `./bin/server.sh`
* accessing http://127.0.0.1:11222/console/ gives "resource not found"

**Comment by pruivo** [comment]
I guess this piece of code from `RestServerConfigurationParser` is missing.
```
if (endpoint.admin()) {
         String serverHome = reader.getProperties().getProperty(Server.INFINISPAN_SERVER_HOME_PATH);
         builder.staticResources(Paths.get(serverHome, Server.DEFAULT_SERVER_STATIC_DIR));
      }
      builder.authentication().metricsAuth(endpoint.metricsAuth());
```

**Comment by tristantarrant** [comment]
@pruivo I've addressed the console issue
@ryanemerson you can now set the security realm individually on the connectors and it will automatically enable all relevant security bits (same as for endpoint):

```
<endpoints socket-binding="default" security-realm="default"/>
```
Enables both HR + REST and secures them.

```
<endpoints socket-binding="default">
    <hotrod-connector  security-realm="default"/>
    <rest-connector/>
</endpoints>
```
Secures the Hot Rod endpoint.

File: server/runtime/src/main/java/org/infinispan/server/configuration/ServerConfigurationParser.java [hunk 2]
```diff
@@ -1299,9 +1303,22 @@ private void parseEndpoints(XMLExtendedStreamReader reader, ConfigurationBuilder
       while (reader.hasNext() && (reader.nextTag() != XMLStreamConstants.END_ELEMENT)) {
          reader.handleAny(holder);
       }
+      if (endpoint.connectors().isEmpty()) {
+         endpoint.addConnector(HotRodServerConfigurationBuilder.class).name("hotrod-" + socketBinding);
+         RestServerConfigurationBuilder rest = endpoint.addConnector(RestServerConfigurationBuilder.class).name("rest-" + socketBinding);
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/configuration/ServerConfigurationParser.java:1308]: We should abstract the creation of the default Hot Rod and REST endpoint names to a static method in the `HotRodServerConfigurationParser` and `RestServerConfigurationParser`.

File: server/runtime/src/main/resources/schema/infinispan-server-12.1.xsd [hunk 3]
```diff
@@ -422,6 +427,11 @@
                   </xs:documentation>
                </xs:annotation>
             </xs:attribute>
+            <xs:attribute name="security-realm" type="xs:string">
+               <xs:annotation>
+                  <xs:documentation>Names the security realm to use for authentication, cache authorization, and encryption. </xs:documentation>
+               </xs:annotation>
+            </xs:attribute>
```
Comment (ryanemerson) [inline server/runtime/src/main/resources/schema/infinispan-server-12.1.xsd:434]: I think we should deprecate the "security-realm" attribute on the HR and REST authentication elements.

**Review by ryanemerson — COMMENTED** [review]
Works as expected for me. Just a couple of minor points.

File: server/runtime/src/main/java/org/infinispan/server/configuration/ServerConfigurationParser.java (hunk 2 above)
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/configuration/ServerConfigurationParser.java:1308]: Agreed.

File: server/runtime/src/main/resources/schema/infinispan-server-12.1.xsd (hunk 3 above)
Comment (tristantarrant) [reply server/runtime/src/main/resources/schema/infinispan-server-12.1.xsd:434]: Yes

**Comment by ryanemerson** [comment]
@tristantarrant The `parseAuthentication` methods need updating for both endpoint parsers so that we check if the `security-realm` has been set on the protocol element. Currently the following throws `ISPN080021: Authentication cannot be configured without a security realm`

```xml
        <endpoints socket-binding="default">
            <hotrod-connector security-realm="default">
                <authentication>
                    <sasl qop="auth" server-name="infinispan"/>
                </authentication>
            </hotrod-connector>

            <rest-connector security-realm="default">
            </rest-connector>

        </endpoints>
```

**Comment by tristantarrant** [comment]
@ryanemerson I've addressed the issues.

**Review by ryanemerson — APPROVED** [review]
The changes work well with the image, nothing more to add from my side. @pruivo anything else to add?

**Comment by pruivo** [comment]
@ryanemerson if everything works, I'm good with it.

**Comment by ryanemerson** [comment]
Thanks @tristantarrant

**Pull request merged**